File: Cpp/AlgorithmComparison.cpp
/**
* Compares various sorting algorithms.
* To execute, compile files using:
* g++ *.cpp -o output
* Run:
* output.exe  
*/

#include <bits/stdc++.h> 
#include <algorithm> 
#include <chrono>
#include<string>

#include "sort_type.h"
#include "sorting_algorithms.h"

using namespace std;

int SIZE = 600;
int MAX_ELEMENT = 1000;

void printSortingTime(SortType sortType, int arr[]){
	int arrCopy[SIZE];
	copy(arr, arr + SIZE, arrCopy);
	string algorithmUsed = "";
	
	auto start = chrono::high_resolution_clock::now();

	switch(sortType){
		case BUBBLE:
			bubbleSort(arrCopy, SIZE);
			algorithmUsed = "BUBBLE_SORT";
			break;
		case COUNT:
			countSort(arrCopy, SIZE);
			algorithmUsed = "COUNT_SORT";
			break;
		case HEAP:
			heapSort(arrCopy, SIZE);
			algorithmUsed = "HEAP_SORT";
			break;
		case INSERTION:
			insertionSort(arrCopy, SIZE);
			algorithmUsed = "INSERTION_SORT";
			break;
		case MERGE:
			MergeSort(arrCopy, 0, SIZE - 1);
			algorithmUsed = "MERGE_SORT";
			break;
		case QUICK:
			quickSort(arrCopy, 0, SIZE - 1);
			algorithmUsed = "QUICK_SORT";
			break;
		case SELECTION:
			selectionSort(arrCopy, SIZE);
			algorithmUsed = "SELECTION_SORT";
			break;
	}

	auto finish = chrono::high_resolution_clock::now();
	auto timeTaken = chrono::duration_cast<chrono::nanoseconds>(finish-start).count();
	printf("%-17s %15d \n", algorithmUsed.c_str(), timeTaken);

}

void compareSortingAlgorithms(int arr[]){
	printSortingTime(SortType::BUBBLE, arr);
	printSortingTime(SortType::COUNT, arr);
    printSortingTime(SortType::HEAP, arr);
    printSortingTime(SortType::INSERTION, arr);
    printSortingTime(SortType::MERGE, arr);
    printSortingTime(SortType::QUICK, arr);
    printSortingTime(SortType::SELECTION, arr);
}

int main(){
	int arr[SIZE];

	for(int i = 0; i < SIZE; i++){
		arr[i] = rand() % MAX_ELEMENT;
	}

	cout << "Array length = " << SIZE << endl;
	cout << "\nElements randomly distributed: " << endl;
	cout << "\nSorting Algorithm\tTime Taken(ns)" << endl;

	compareSortingAlgorithms(arr);
	
	cout << "\nElements already sorted: " << endl;
	cout << "\nSorting Algorithm\tTime Taken(ns)" << endl;
	
	sort(arr, arr + SIZE);
	compareSortingAlgorithms(arr);

	cout << "\nElements sorted in reverse order:" << endl;
	cout << "\nSorting Algorithm\tTime Taken(ns)" << endl;

	reverse(arr, arr + SIZE);
	compareSortingAlgorithms(arr);
}

File: Go/AlgorithmComparison.go
// To execute this source file, all the listed sorting algorithms' files must be placed within the same
// directory as this one. There should be no other main function present in any of the listed algorithm files.
// If present, comment those main functions in order to execute this file.

// Additionally, the sort functions should all have unique names and the same signature, i.e.
// func([]int, int) []int {}

// To execute, use the command "go run AlgorithmComparison.go a1.go a2.go a3.go ... an.go", where a1, a2,
// a3, ..., an represents the names of the sorting algorithm files as present in the directory.

package main

import (
	"fmt"
	"math/rand"
	"sort"
	"time"
)

var algorithms = map[string]interface{}{
	"Bitonic Sort":   BitonicSort,
	"Bubble Sort":    BubbleSort,
	"Bucket Sort":    BucketSort,
	"Comb Sort":      CombSort,
	"Cycle Sort":     CycleSort,
	"Gnome Sort":     GnomeSort,
	"Heap Sort":      HeapSort,
	"Insertion Sort": InsertionSort,
	"Quick Sort":     QuickSort,
	"Radix Sort":     RadixSort,
	"Selection Sort": SelectionSort,
	"Shell Sort":     ShellSort,
	"Tim Sort":       TimSort,
}

func timeTaken(algorithm string, arr []int) int64 {
	startTime := time.Now()
	algorithms[algorithm].(func([]int, int) []int)(arr, len(arr))
	timeDuration := time.Since(startTime)

	totalTime, _ := time.ParseDuration(timeDuration.String())

	return totalTime.Nanoseconds()
}

func main() {
	var arr []int
	n := 6000

	fmt.Printf("\nArray length = %d\n", n)
	fmt.Println("\nElements randomly distributed : ")
	fmt.Println("\nSorting Algorithm\tTime Taken(ns)")
	for k, _ := range algorithms {
		arr = rand.Perm(n)

		fmt.Println(k, "\t\t", timeTaken(k, arr))
	}

	fmt.Println("\nElements already sorted : ")
	fmt.Println("\nSorting Algorithm\tTime Taken(ns)")
	for k, _ := range algorithms {
		arr = rand.Perm(n)
		sort.Ints(arr)

		fmt.Println(k, "\t\t", timeTaken(k, arr))
	}

	fmt.Println("\nElements sorted in reverse order : ")
	fmt.Println("\nSorting Algorithm\tTime Taken(ns)")
	for k, _ := range algorithms {
		arr = rand.Perm(n)
		sort.Ints(arr)
		sort.Sort(sort.Reverse(sort.IntSlice(arr)))

		fmt.Println(k, "\t\t", timeTaken(k, arr))
	}
}
